New tutorial for NodeJS & TypeScript

dotansimha opened the pull request:
Hi there!

This PR adds a new tutorial for creating GraphQL servers using the following:

- Node.js
- TypeScript
- fastify
- graphql-js
- graphql-helix
- Prisma 2
- GraphQL Playground

It follows almost the same structure and wording of the existing node-graphql tutorial but it’s up to date and explains the core concepts in a better way.

It is also detached from any framework and focuses on GraphQL and will work with any HTTP framework (we demonstrated here using Fastify)

We would love your review and getting this merged as we want to direct a lot of people to that, we think it’s really missing right now for everyone in the ecosystem.

We also started working on an advanced tutorial and we’ll send a PR once that will be merged.

The complete implemented server is located here: https://github.com/dotansimha/graphql-typescript-node-tutorial (it based on steps, so each step in the tutorial have it's own commit in the repo, and I would love to move it under the `howtographql` org if needed)

dotansimha commented:
Hi, @TasinIshmam, nice to meet you, and thank you so much for your comment!

1. I'll review the guidelines and update the content, thank you!
1. About the name - I don't mind changing the name but an important point about this tutorial is that it is deliberately structured to be HTTP framework agnostic.  So I'm not sure naming it `typescript + fastify` or `typescript + helix` would give the right context. I think the way we are having it today, missed the point of separating these different concepts. So I've created this tutorial and the chapters in a way that I would explain the separation in a way that we won't need to have another tutorial for express
1. For the same point and separation - because of that separation, we won't need another tutorial for let's say Nexus, TypeGraphQL or other schema builders. all of these could become part of the same tutorial and people would be aware of all choices and choose their path.
1. Because of that - what name would you give this tutorial?
1. On the same subject - could you share the tutorial or structure you are creating in Prisma?  maybe there is a way to merge the two together? (if let's say the main difference is the library choices like express and Nexus for example).  That way we could help you finish it, maintain it and it would also be less confusing to learners.
1. I'm going to write here on Wednesday the outline of the advance tutorial, would love to get your feedback on it!
1. About the timeline for merging, we have a lot of need from our community to have it up, are you ok with merging it before and later we could address all your comments and improve any feedback you might have? if not that's ok and we'll like our community to the preview link for now :)

Let me know what you think :)

TasinIshmam commented:
Hey @dotansimha, thanks for the thorough feedback.

> 1. I'll review the guidelines and update the content, thank you!

Thanks, great to know :smile:  

> 2. About the name - I don't mind changing the name but an important point about this tutorial is that it is deliberately structured to be HTTP framework agnostic.  So I'm not sure naming it `typescript + fastify` or `typescript + helix` would give the right context. I think the way we are having it today, missed the point of separating these different concepts. So I've created this tutorial and the chapters in a way that I would explain the separation in a way that we won't need to have another tutorial for express

I understand your point about keeping the "conceptual" and "implementation" materials separated. I would say the separation already exists on the website as there are quite a few articles that discuss graphql at a conceptual level, detached from any specific framework or language. 

Nevertheless, most of the other tutorials go into the conceptual as well by providing a small introduction to the topic or linking to relevant resources. However, the main objective of these tutorial tracks is for a new user (of any skill level) to learn how to _implement_ a GraphQL server/client. So, the tutorials should have self-explanatory titles and descriptions so that a user can look at them to immediately understand what tools/frameworks will be used to do the actual implementation. This is especially true for tutorials where there are multiple alternatives (such as the react one), so the user's can make the best choice given their preferences and experience level. 

> 3. For the same point and separation - because of that separation, we won't need another tutorial for let's say Nexus, TypeGraphQL or other schema builders. all of these could become part of the same tutorial and people would be aware of all choices and choose their path.


I think a tutorial should only claim to be covering the technology(ies) that it explicitly uses. Using the example you mention, if a tutorial _uses_ Nexus and Fastify to implement/demonstrates various GraphQL concepts, I think it would be unreasonable to claim that the tutorial has also "covered" Express and TypeGraphQL (even if those implementations are similar). 

> 4. Because of that - what name would you give this tutorial?

I hope you have a better understanding of my viewpoint based on the last few paragraphs :sweat_smile:. I would encourage specific names tied to easily understandable technologies or paradigms, instead of something general like ```GraphQL HTTP```. 

Mainly based on the explanation given in the [introduction](https://github.com/howtographql/howtographql/blob/c2e914bec655d9e678861a8b814c488f2544b1db/content/backend/graphql-node-typescript/0-introduction.md), I still think helix or fastify are the most _representative_ parts of the technology stack. I would love more suggestions/options on this if you have any. 




> 5. On the same subject - could you share the tutorial or structure you are creating in Prisma? 

Sure, I'll write up and send you a general outline once I get back.

> maybe there is a way to merge the two together? (if let's say the main difference is the library choices like express and Nexus for example).  That way we could help you finish it, maintain it and it would also be less confusing to learners.

I'm not sure if we could feasibly merge the two tutorials given that they are both in a pretty late stage/almost finished :disappointed: . We discussed this internally and we feel that it would be easier to keep them separate for the moment. 

However, we'd love to collaborate on launch and promotion. Additionally, If you have any ideas for something completely new to work on together, that's something we'd love to discuss.

> 6. I'm going to write here on Wednesday the outline of the advanced tutorial, would love to get your feedback on it!

Sure, that sounds great! Please create a new issue for this so we can keep the two topics/discussions separate.


> 7. About the timeline for merging, we have a lot of need from our community to have it up, are you ok with merging it before and later we could address all your comments and improve any feedback you might have? if not that's ok and we'll like our community to the preview link for now :)

If possible, please use the preview link for now. I would really prefer to ensure the series is properly polished before publishing it to the website. I hope you understand. 



Btw a few other issues (non-exhaustive) that I noticed at a glance

1. There are two sections titled "Getting Started". 
2. A lot of the code is missing proper syntax highlighting
3. Most of the code is missing change highlighting (Which lines have been modified from a previous code snippet) 
4. There's an MCQ question that accompanies each chapter as a concept refresher. I would strongly recommend adding these. 
5. Code snippet lack path argument (for filename) 
6. Split up ```schema.ts``` into multiple files. 


Sorry I haven't been able to respond very promptly. I look forward to looking at plans for the advanced tutorial you mentioned :ok_hand:. I will try my best to get back to you as soon as I can.

saihaj commented on content/backend/graphql-node-typescript/8-authentication.md:
```diff
@@ -0,0 +1,522 @@
+contextFactory: () => {
+  let currentUser = null;
+
+  if (req.headers.authorization) {
+    // 1
+    const token = req.headers.authorization.split(" ")[1];
+    // 2
+    const tokenPayload = verify(token, APP_SECRET) as JwtPayload;
+    // 3
+    const userId = tokenPayload.userId;
+    // 4
+    currentUser = prisma.user.findUnique({ where: { id: userId } });
+  }
```
Right now if token is invalid or empty request doesn't go through and user cannot access login mutation. 
```suggestion
  if (req.headers.authorization) {
    // 1
    const token = req.headers.authorization.split(" ")[1];
    try {
      // 2
      const tokenPayload = verify(token, APP_SECRET) as JwtPayload;
      // 3
      const userId = tokenPayload.userId;
      // 4
      currentUser = prisma.user.findUnique({ where: { id: userId } });
    } catch {
      // We allow the request to go through so user can log in or access other unauthenticated operations
    }
  }
```

saihaj commented on content/backend/graphql-node-typescript/8-authentication.md:
````diff
@@ -0,0 +1,522 @@
+Then, let's modify the `contextFactory` in `src/index.ts` function to do this flow:
+
+```ts
+contextFactory: () => {
+  let currentUser = null;
+
+  if (req.headers.authorization) {
+    // 1
+    const token = req.headers.authorization.split(" ")[1];
+    // 2
+    const tokenPayload = verify(token, APP_SECRET) as JwtPayload;
+    // 3
+    const userId = tokenPayload.userId;
+    // 4
+    currentUser = prisma.user.findUnique({ where: { id: userId } });
````
```suggestion
contextFactory: async () => {
  let currentUser = null;

  if (req.headers.authorization) {
    // 1
    const token = req.headers.authorization.split(" ")[1];
    // 2
    const tokenPayload = verify(token, APP_SECRET) as JwtPayload;
    // 3
    const userId = tokenPayload.userId;
    // 4
    currentUser = await prisma.user.findUnique({ where: { id: userId } });
```

dotansimha commented:
Hi @TasinIshmam , I pushed fixes here and to the repo, mainly:

- Fixed `we/you` all across the tutorial
- Migrated from Playground to GraphiQL (there is work-in-progress to merge the two, and the latest GraphiQL supports SSE and all popular transports)
- Added a chapter on Subscriptions with SSE (simpler than WebSocket, easier to integrate and teach)
- Refactored some pieces of the codebase, simplified (some parts) of `schema.ts` 
- Updated all screenshots.
- Updated all code examples, missing paths, and missing line highlights. 

Still missing:
- Quiz questions for all chapters
- Code Review

Regarding the name of the tutorial, I wanted to suggest a different take: What if we'll consider omitting the parts the are easily replaceable? (For example, TS and NodeJS and the core of the tutorial, `fastify` could be easily replaced with any HTTP server, and GraphQL-Helix is a core)
This will lead us to `nodejs + typescript + helix` which makes a lot of sense, at least for me.

TasinIshmam commented:
Thanks for the update @dotansimha ! 

Let me take a look through the whole tutorial and get back to you.

yograf commented:
Great job @dotansimha, thanks.

TasinIshmam commented:
Hi, I'm extremely sorry for the delay, I have started reviewing the content today and will be finished over the weekend.

TasinIshmam commented on content/backend/graphql-node-typescript/3-graphql-server.md:
```diff
@@ -0,0 +1,242 @@
+1. The `processRequest` gets all the input parameters, and runs the GraphQL request pipeline for us.
+1. The result of the GraphQL operation is being returned and sent to the client.  
+
+> GraphQL-Helix supports real-time GraphQL and stream-based responses, that's why your code need to check it's result before sending it. At the moment, our server only supports regular responses (`type="RESPONSE"`). You'll later add support for more transports.  
```
This sentence introduces a lot of concepts (stream-based responses, transports) that a beginner is unlikely to be familiar with. Please simply this sentence, explain the necessary concepts and provide appropriate resources where necessary.

```suggestion
> GraphQL-Helix supports real-time GraphQL and stream-based responses, that's why your code needs to check the result before sending it. At the moment, our server only supports regular responses (`type="RESPONSE"`). You'll later add support for more transports.  
```

TasinIshmam commented on content/backend/graphql-node-typescript/3-graphql-server.md:
```diff
@@ -0,0 +1,242 @@
+
+A complete GraphQL flow is consists of the following steps:
+
+1. Schema building - first, make sure that you have a GraphQL schema ready-to-use (built as executable schema). 
```
This seems like a warning or instruction whereas it should be a step in the GraphQL Flow.

TasinIshmam commented on content/backend/graphql-node-typescript/3-graphql-server.md:
```diff
@@ -0,0 +1,242 @@
+1. Schema building - first, make sure that you have a GraphQL schema ready-to-use (built as executable schema). 
+1. `parse` - takes the GraphQL operation and parse it into a `DocuemntNode`.
+1. `validate` - uses the parsed GraphQL operation with the GraphQL schema, and validate it to make sure it matches the schema.
+1. Variables - GraphQL operations allow you to pass variables, which are just JSON object. 
+1. Context building - GraphQL allow you to pass a custom object, called `context`, and this object will be available for your resolvers. 
```
Are these steps in the GraphQL flow? The way it is written, they seem more like features supported by GraphQL.

TasinIshmam commented on content/backend/graphql-node-typescript/3-graphql-server.md:
```diff
@@ -0,0 +1,242 @@
+1. Context building - GraphQL allow you to pass a custom object, called `context`, and this object will be available for your resolvers. 
+1. `execute` - takes the operation and runs the resovlers based on the selection-set.
+
+With `graphql-helix`, this flow becomes simpler because you can just use `processRequest`.
```
Why does it become easier? What `processRequest` actually does is not clarified.

TasinIshmam commented on content/backend/graphql-node-typescript/3-graphql-server.md:
```diff
@@ -0,0 +1,242 @@
+
+</Instruction>
+
+Now, run the server, and try to access `http://localhost:3000/graphql` in your browser. You should see GraphiQL UI, and from there you can explore the schema, the documentation, and even try to run queries! 
```
You should provide a screenshot of what the user should expect to see. Furthermore, you should explain some of the "basics" of the graphiql UI so the user understands what they are looking at.

TasinIshmam reviewed (changes requested):
Feedback for parts 0, 1, 2 and 3.

dotansimha commented:
Thanks @TasinIshmam for the review, this was really helpful!! I applied all the changes you suggested and added the missing screenshots. Also, I updated the latest helix and now the code is simpler, so no need to dive deep into explaining streaming and so on.

TasinIshmam commented:
> Thanks @TasinIshmam for the review, this was really helpful!! I applied all the changes you suggested and added the missing screenshots. Also, I updated the latest helix and now the code is simpler, so no need to dive deep into explaining streaming and so on.

Thank you! Reviewing the rest of the tutorial right now and will update that soon.

TasinIshmam commented:
Hey @dotansimha  I have finished reviewing all the content, please take a look :sweat_smile:  


### Naming

> Regarding the name of the tutorial, I wanted to suggest a different take: What if we'll consider omitting the parts the are easily replaceable? (For example, TS and NodeJS and the core of the tutorial, `fastify` could be easily replaced with any HTTP server, and GraphQL-Helix is a core) This will lead us to `nodejs + typescript + helix` which makes a lot of sense, at least for me.

Would you be alright with considering `typescript + helix`. The `nodejs` convention that we have now is a bit redundant and makes the name verbose. Most people would be easily able to discern that this is a backend (hence nodejs) tutorial. We were actually planning to take it out of the other tutorial(s) in the JS/TS family as well.

saihaj commented:
Hey @TasinIshmam sorry for the delay on our end. I have addressed all the comments please review it and let us know if something else is needed. Looking at other examples I think we need to get  https://github.com/dotansimha/graphql-typescript-node-tutorial under this Org so that code examples are linked. Let us know how to proceed with this. Thanks!

TasinIshmam commented:
I took a quick look and I think everything is fine and we can merge the PR! Thanks for the great work @saihaj  and @dotansimha . 

I made a change to the question in chapter 4, because it was a bit ambiguous and not in the correct format, hope that's okay. 


### Transfering the repo 

To transfer a repo to an organization, one needs permission to create a repository in an organization. So please do the following regarding the transfer of [the code repo](https://github.com/dotansimha/graphql-typescript-node-tutorial). 

1. If any additional changes need to be made so that the repo matches the code in the tutorial, please do so.
2. Rename the repository to `typescript-helix`  
3. Transfer it to my GitHub user account (@tasinishmam)  (or I could fork the repo as well. If you would prefer that, let me know).

Afterward, I will transfer it to the howtographql org and give you both (@dotansimha  and @saihaj) edit access to the code repo. 

Btw I see that the code repo uses `master` and not `main`. _Please keep it this way and don't change it  as the website looks for the master branch when linking any code._

dotansimha commented:
> I took a quick look and I think everything is fine and we can merge the PR! Thanks for the great work @saihaj and @dotansimha .
> 
> I made a change to the question in chapter 4, because it was a bit ambiguous and not in the correct format, hope that's okay.

Great, thank you!

> ### Transfering the repo
> To transfer a repo to an organization, one needs permission to create a repository in an organization. So please do the following regarding the transfer of [the code repo](https://github.com/dotansimha/graphql-typescript-node-tutorial).
> 
> 1. If any additional changes need to be made so that the repo matches the code in the tutorial, please do so.
> 2. Rename the repository to `typescript-helix`
> 3. Transfer it to my GitHub user account (@TasinIshmam)  (or I could fork the repo as well. If you would prefer that, let me know).
> 
> Afterward, I will transfer it to the howtographql org and give you both (@dotansimha and @saihaj) edit access to the code repo.
> 
> Btw I see that the code repo uses `master` and not `main`. _Please keep it this way and don't change it as the website looks for the master branch when linking any code._

Done, you should get an email about the transfer :) Thanks!

TasinIshmam commented:
PR has been merged and the code repository has been transferred :tada:  

@dotansimha  and @saihaj, you should both have write-access to the code repository if any changes are necessary. If you would like to make any changes to the tutorial itself, please make a PR and I will be happy to take a look at it.

Thanks again for your contributions!
